cli/src/cmd/forge/script/runner.rs

```diff
@@ -278,14 +278,77 @@ impl ScriptRunner {
 
         let mut res = self.executor.call_raw(from, to, calldata.0.clone(), value)?;
         let mut gas_used = res.gas_used;
+
+        // We should only need to calculate realistic gas costs when preparing to broadcast
+        // something. This happens during the onchain simulation stage, where we commit each
+        // collected transactions.
+        //
+        // Otherwise don't re-execute, or some usecases might be broken: https://github.com/foundry-rs/foundry/issues/3921
+        if commit {
+            gas_used = self.search_optimal_gas_usage(&res, from, to, &calldata, value)?;
+
+            // if we changed `fs_commit` during gas limit search, re-execute the call with original
+            // value
+            if fs_commit_changed {
+                if let Some(ref mut cheatcodes) = self.executor.inspector_config_mut().cheatcodes {
+                    cheatcodes.fs_commit = !cheatcodes.fs_commit;
+                }
+            }
```

Comment:
I'm unsure of this... @mattsse

Reply:
this sounds like a very unlikely edge case? but hard to say without an example.